MAINT: signal.csd: port away from using `_spectral_helper`

DietBru opened the pull request:
This allows to mark the internal functions `_spectral_helper`, `fft_helper` and `_triage_segements` as legacy, since they are only used by the legacy functions `spectrogram` and `stft` now. Further changes:

* Some small additions to the unit tests to achieve 100% coverage for `csd()`.
* Remove csd related functions from file `_scipy_spectral_tst_shim.py`
* An explanation of the `ShortTimeFFT` usage was added to the `csd()` docstr.

### Related Issues and PRs:

#12544: Close as wont fix, since only `_triage_segements` is affected.
#13577: Close as wont fix, since only `_spectral_helper` is affected . Furthermore, `ShortTimeFFT` supports "chunking".
#15593: Close, since only `_triage_segements` is changed.
#15587: Close as wont fix, since only `_triage_segements` is affected.
#19001: Edits relevant parts of `spectral_py.py`


~~Closes~~ ~~#22674 (check comment).~~

DietBru committed:
MAINT: signal.csd(): Port away from using `_spectral_helper`

* This allows to mark the internal functions `_spectral_helper`, `fft_helper` and `_triage_segements` as legacy, since they are only used by the legacy functions `spectrogram` and `stft`.

* Some small additions to the unit tests to achieve 100% coverage for `csd()`.

* Remove csd related functions from file `_scipy_spectral_tst_shim.py`

* An explanation of the `ShortTimeFFT` usage was added to the `csd()` docstr.

j-bowhay commented on scipy/signal/_spectral_py.py:
```diff
@@ -759,58 +761,165 @@ def csd(x, y, fs=1.0, window='hann', nperseg=None, noverlap=None, nfft=None,
+        x_outer.pop(axis)
+        y_outer.pop(axis)
+        try:
+            outer_shape = np.broadcast(np.empty(x_outer), np.empty(y_outer)).shape
```
We should use `np.broadcast_shapes` here

DietBru replied:
Good point. Fixed in Commit [04c8afd](https://github.com/scipy/scipy/pull/22460/commits/04c8afdb890013f421ef7cc0a3ffb2bcda7b6262).

DietBru committed:
signal.csd(): Changed `np.broadcast` into the more efficient`np.broadcast_shapes`

DietBru committed:
Improved docstrs of `signal.welch()` and `signal.csd()` by clarifying relationship between squared magnitude/cross spectrum and PSD/cross PSD.

DietBru committed:
Improved docstrs of `signal.periodogram()`  by clarifying relationship between squared magnitude/cross spectrum and PSD/cross PSD.

DietBru commented:
Rendered docs:
* [csd](https://output.circle-artifacts.com/output/job/2b0345cf-17e8-4b34-8beb-8ff74b2cda2f/artifacts/0/html/reference/generated/scipy.signal.csd.html)
* [welch](https://output.circle-artifacts.com/output/job/2b0345cf-17e8-4b34-8beb-8ff74b2cda2f/artifacts/0/html/reference/generated/scipy.signal.welch.html)
* [periodogram](https://output.circle-artifacts.com/output/job/2b0345cf-17e8-4b34-8beb-8ff74b2cda2f/artifacts/0/html/reference/generated/scipy.signal.periodogram.html#scipy.signal.periodogram)

ev-br commented:
It must be frustrating to have a ready PR just sit there for more than a month. Sorry about that @DietBru !
To help moving this forward, could you please give a high-level overview of this spectral analysis functionality: what is the current state and what is the direction it needs moving?

DietBru commented:
> It must be frustrating to have a ready PR just sit there for more than a month. Sorry about that @DietBru ! To help moving this forward, could you please give a high-level overview of this spectral analysis functionality: what is the current state and what is the direction it needs moving?

Well, my available time for working on SciPy is also quite unevenly distributed (as you probably noted)—so I have the same problem of not being able to review PRs as much as I would like.

First to the rationale of this PR: The class `ShortTimeFFT` provides a superset of functionality of `_spectral_helper`. This illustrated by the shims provided in the file [_scipy_spectral_test_shim.py]. The aim of this PR is port all non-legacy functions away from `_spectral_helper`, since it has a couple of bugs (as shown above). It would reduce the amount code and make porting to the Array API easier. A next step would be discussing, which legacy functions could be deprecated (i.e, [spectrogram], [stft] and [istft]).

[_scipy_spectral_test_shim.py]: https://github.com/scipy/scipy/blob/main/scipy/signal/tests/_scipy_spectral_test_shim.py
[spectrogram]: https://scipy.github.io/devdocs/reference/generated/scipy.signal.spectrogram.html
[stft]: https://scipy.github.io/devdocs/reference/generated/scipy.signal.stft.html#scipy.signal.stft
[istft]: https://scipy.github.io/devdocs/reference/generated/scipy.signal.istft.html#scipy.signal.istft

Concerning the spectral analysis functionality, I'll just discuss calculating (cross-) spectral power density (PSD) aspects here: From an algorithmic view, the signal module is not in bad shape. Using them properly is not straightforward though, since the choice of parameters can change the underlying modelling assumptions significantly. Hence, I think the main challenge lies in providing useful guidance in the documentation.

To elaborate: The classical Welch algorithm view assumes a stationary stochastic process and provides little insights on choosing window and overlap. This PR illustrates (again) that the Welch algorithm is equivalent to determining the frequency marginal of a `ShortTimeFFT` spectrogram (assuming the  window is correctly scaled). The introduction of dual windows for the `ShortTimeFFT` allows to interpret the STFT as a deterministic signal decomposition into a non-orthogonal over-complete series expansion (aka frame). Hence, the Welch algorithm calculates a statistical property of that series expansion, which, obviously, depends on the choice of (dual) window. This leads to the very unsatisfying conclusion that the PSD inherently depends on the choice of parameters.

For many practical applications, it is sufficient to calculate a reasonable PSD, which allows comparing signals to each other—hence the normal heuristics suffice. On the other hand, when interpreting spectral cross-densities,  the shape of the dual window isn't always negligible. Unfortunately, I am not aware of any literature, which discusses such problems for practical applications. A chapter or two in a textbook is probably more suited for discussing these kind of modeling problems than the SciPy documentation. Sorry, for not being able to provide a more useful answer here.

ev-br commented:
Thanks @DietBru! This is a really helpful summary. I'm not claiming detailed knowledge or understanding of details here, and I'm not going to ask you for a couple of chapters worth of textbook for the SciPy docs :-). My impression is that the direction makes total sense, the PR is mostly maintenance + doc improvements with no hard backwards compat breaks and a clear path forward.  
So from my POV this PR is good to go in as is. 

That said, if you want to take a closer look @larsoner, I happily leave it all to you.

DietBru committed:
Small Improvement to docstr fo `signal._spectral_helper`

DietBru commented on scipy/signal/_spectral_py.py:
```diff
@@ -1890,6 +2016,11 @@ def _spectral_helper(x, y, fs=1.0, window='hann', nperseg=None, noverlap=None,
+    .. legacy:: function
+
+        This function is soley used by the legacy functions `spectrogram` and `stft`
+        (which are also in this file).
```
`_spectral_helper` is a private function—so no HTML should be generated. Add  the file in commit 5b00e150b30765b6f79df8472ceabf24407cf042, since it doesn't hurt.

DietBru commented on scipy/signal/_spectral_py.py:
```diff
@@ -759,58 +778,165 @@ def csd(x, y, fs=1.0, window='hann', nperseg=None, noverlap=None, nfft=None,
+
+    # using cast() to make mypy happy:
+    fft_mode = cast(FFT_MODE_TYPE, 'onesided' if return_onesided else 'twosided')
+    if scaling not in (scales := {'spectrum': 'magnitude', 'density': 'psd'}):
```
﹕:⋮ :smile: ⋮:﹕

tylerjereddy reviewed (commented):
I think I'm on the same page as Evgeni here--in favor of moving forward with this PR before we branch.

Also not a domain expert, but added a few comments in line.

Rebased version of this branch is still passing full suite locally.

larsoner reviewed (approved):
LGTM after addressing @tylerjereddy's comments!

DietBru committed:
ENH `signal.csd`: Incorporate review comments to docstr of `signal.csd`.

[docs only]

DietBru commented:
Thanks for the input, @tylerjereddy. I addressed your remarks in commit [11c0c45](https://github.com/scipy/scipy/pull/22460/commits/11c0c45be6439b5fa8d993c1847b1388973973b2).

tylerjereddy commented:
Looks like my docs-related comments were addressed and there are three core devs in favor of a merge so in it goes, thanks.
